autodiscern/annotations.py

```diff
@@ -0,0 +1,237 @@
+from allennlp.data.tokenizers.word_tokenizer import WordTokenizer
+from allennlp.predictors.predictor import Predictor
+from bs4 import BeautifulSoup
+from bs4.element import Tag
+import re
+from typing import Callable, Dict, List, Tuple
+from autodiscern.data_manager import DataManager
+
+
+def add_word_token_annotations(inputs: Dict[str, Dict]) -> Dict[str, Dict]:
+    tok = WordTokenizer()
+    for id in inputs:
+        # have to convert tokens to text because spacy tokens are not pickleable
+        inputs[id]['tokens'] = [t.text for t in tok.tokenize(inputs[id]['content'])]
+    return inputs
+
+
+def add_metamap_annotations(inputs: Dict[str, Dict], dm: DataManager, metamap_path: str = None) -> Dict[str, Dict]:
+    from pymetamap import MetaMapLite
+
+    if metamap_path is None:
+        print("NOTE: no Metamap path provided. Using Laura's default")
+        metamap_path = '/Users/laurakinkead/Documents/metamap/public_mm_lite/'
+
+    metamap_semantics = dm.metamap_semantics
+    groups_to_keep = [
+        'Anatomy',
+        'Devices',
+        'Disorders',
+        'Physiology',
+        'Procedures',
+    ]
+    types_to_keep = [
+        # Chemicals & Drugs
+        'Antibiotic',
+        'Clincal Drug',
+        'Enzyme',
+        'Hormone',
+        'Pharmacologic Substance',
+        'Receptor',
+    ]
+    semantic_type_filter_df = metamap_semantics[metamap_semantics['group_name'].isin(groups_to_keep) |
+                                                metamap_semantics['name'].isin(types_to_keep)
+                                                ].sort_values(['group_name', 'name'])
+    semantic_type_filter = list(semantic_type_filter_df['abbreviation'])
+
+    # create list of ids, and list of sentences in matching order
+    ids = inputs.keys()
+    sentences = []
+    for id in ids:
+        sentences.append(inputs[id]['content'])
+
+    # run metamap
+    print("Extracing MetaMap concepts...")
+    mm = MetaMapLite.get_instance(metamap_path)
+    concepts, error = mm.extract_concepts(sentences, ids)
+
+    print("Attaching Metamap concepts...")
+    # add concepts with score above 1 and matching the semantics filter to input sentences
```

Comment:
As we discussed in our last meeting, we might need to move the metamap annotation to the word level (i.e. create a binary vector with length equal to the high-level categories that keeps count of the metamap concepts annotation). Or a one-hot encoded vector that keeps track of metamap concept with highest score.

Reply:
I've created a To Do task for this: #37 
I'm not going to include it in this PR

Reply:
OK.